src/utils.zig

```diff
@@ -1530,6 +1530,228 @@ pub const BinaryWriter = struct {
+		ReadWriteTest.TestEnum(u16),
+		ReadWriteTest.TestEnum(u32),
+		ReadWriteTest.TestEnum(i2),
+		ReadWriteTest.TestEnum(i4),
+		ReadWriteTest.TestEnum(i5),
+		ReadWriteTest.TestEnum(i8),
+		ReadWriteTest.TestEnum(i16),
+		ReadWriteTest.TestEnum(i32),
+	}) |enumT| {
+		try ReadWriteTest.testEnum(enumT, .first);
+		try ReadWriteTest.testEnum(enumT, .center);
+		try ReadWriteTest.testEnum(enumT, .last);
+	}
+}
+
+test "read/write Vec3i" {
+	ReadWriteTest.init();
+	defer ReadWriteTest.deinit();
+
+	try ReadWriteTest.testVec(main.vec.Vec3i, .{0, 0, 0});
+	try ReadWriteTest.testVec(main.vec.Vec3i, .{
+		std.math.maxInt(@typeInfo(main.vec.Vec3i).vector.child),
+		std.math.minInt(@typeInfo(main.vec.Vec3i).vector.child),
+		std.math.minInt(@typeInfo(main.vec.Vec3i).vector.child),
+	});
+	try ReadWriteTest.testVec(main.vec.Vec3i, .{
+		std.math.minInt(@typeInfo(main.vec.Vec3i).vector.child),
+		std.math.maxInt(@typeInfo(main.vec.Vec3i).vector.child),
+		std.math.maxInt(@typeInfo(main.vec.Vec3i).vector.child),
+	});
+}
+
+test "read/write Vec3f/Vec3d" {
+	ReadWriteTest.init();
+	defer ReadWriteTest.deinit();
+
+	inline for([_]type{main.vec.Vec3f, main.vec.Vec3d}) |vecT| {
+		try ReadWriteTest.testVec(vecT, .{0, 0, 0});
+		try ReadWriteTest.testVec(vecT, .{0.0043, 0.01123, 0.05043});
+		try ReadWriteTest.testVec(vecT, .{5345.0, 42.0, 7854.0});
+		try ReadWriteTest.testVec(vecT, .{
+			std.math.floatMax(@typeInfo(vecT).vector.child),
+			std.math.floatMin(@typeInfo(vecT).vector.child),
+			std.math.floatMin(@typeInfo(vecT).vector.child),
+		});
+		try ReadWriteTest.testVec(vecT, .{
+			std.math.floatMin(@typeInfo(vecT).vector.child),
+			std.math.floatMax(@typeInfo(vecT).vector.child),
+			std.math.floatMax(@typeInfo(vecT).vector.child),
+		});
+	}
+}
+
+test "read/write mixed" {
+	ReadWriteTest.init();
+	defer ReadWriteTest.deinit();
+
+	const type_first = u4;
+	const expected_first = 5;
+
+	const type_second = main.vec.Vec3i;
+	const expected_second = type_second{3, -10, 44};
+
+	const type_third = enum(u3) {first, second, third};
+	const expected_third = .second;
+
+	const type_fourth = f32;
+	const expected_fourth = 0.1234;
+
+	const expected_fifth = "Hello World!";
+
+	var writer = ReadWriteTest.getWriter();
+	defer writer.deinit();
+
+	writer.writeInt(type_first, expected_first);
+	writer.writeVec(type_second, expected_second);
+	writer.writeEnum(type_third, expected_third);
+	writer.writeFloat(type_fourth, expected_fourth);
+	writer.writeSlice(expected_fifth);
+
+	var reader = ReadWriteTest.getReader(writer.data.items);
+
+	try std.testing.expectEqual(expected_first, try reader.readInt(type_first));
+	try std.testing.expectEqual(expected_second, try reader.readVec(type_second));
+	try std.testing.expectEqual(expected_third, try reader.readEnum(type_third));
+	try std.testing.expectEqual(expected_fourth, try reader.readFloat(type_fourth));
+	try std.testing.expectEqualStrings(expected_fifth, try reader.readSlice(expected_fifth.len));
```

Comment:
For the mixed test I would like to also see an is empty assert at the end. For the others I don't mind it, but here it is relevant in my opinion.